src/bisect.cpp

```diff
@@ -132,23 +153,104 @@ std::vector<long> bisect_multilevel(bulk::world& world,
         }
 
 
-        std::map<std::vector<long>, int> edge_counts;
-        std::map<std::vector<long>, long> ids;
-        std::unordered_set<long> remove_nets;
+        if (simplify_mode == "complete") {
+            if (world.rank() == 0) {
+                world.log("Size before par simplify: %d", HC_list[nc_par].global_number_nets());
+            }
 
-        for (auto n : HC_list[nc_par].nets()) {
-            edge_counts[n.vertices()] += n.cost();
-            ids[n.vertices()] = n.id();
-        }
+            // Communicate globally which net ids belong to the local hypergraphs
+            auto netidowner_queue = bulk::queue<int, long>(world);
+            for (auto n : HC_list[nc_par].nets()) {
+                if (n.size() != n.global_size()) {
+                    for (int t = 0u; t < world.active_processors(); t++) {
+                        if (world.rank() != t) {
+                            netidowner_queue(t).send(world.rank(), n.id());
+                        }
+                    }
+                }
+            }
+            world.sync();
+
+            // Communicate pins targeted only to relevant processors
+            auto addvertex_queue = bulk::queue<int, long, long>(world);
+            for (const auto& [sender, netid] : netidowner_queue) {
+                if (HC_list[nc_par].is_local_net(netid)) {
+                    for(auto v : HC_list[nc_par].net(netid).vertices()) {
+                        addvertex_queue(sender).send(world.rank(), v, netid);
+                    }
+                }
+            }
+            world.sync();
+
+            // Add pins locally from the queues
+            for (const auto& [sender, vtoadd, targetnet] : addvertex_queue) {
+                if (HC_list[nc_par].is_local_net(targetnet)) {

```

Comment:
I think this if statement is unnecessary, as you only send this information to a processor that has requested it.